rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,392 @@
+becomes unavailable requiring user intervention.
+
+> However, absolute durability is impossible, because all hardware can
+> ultimately fail. Data we write today might not be available tomorrow.
+> TigerBeetle embraces limited disk reliability and maximizes data durability
+> in spite of imperfect disks. We actively work against such entropy by taking
+> advantage of cluster-wide storage. A record would need to get corrupted on
+> all replicas in a cluster to get lost, and even in that case the system would
+> safely halt.
+
+See https://docs.tigerbeetle.com/about/safety/
+
+#### Pebble
+Pebble is a key-value store inspired by LevelDB and RocksDB, designed for high
+performance and internal use by CockroachDB (CRDB). It is written in Go and
+serves as the physical durability layer for CRDB, while CRDB provides logical
+durability on top of nodes running Pebble.
+
+##### Truncates Faulty WAL Frames
+Similar to SQLite, Pebble truncates all subsequent WAL entries when it detects
+corruption in the WAL. However, unlike SQLite, Pebble logs a warning by
+default.
+
+```
+[JOB 1] WAL file 000002.log with log number 000002 stopped reading at offset: 36289; replayed 869 keys in 869 batches
+```
+
+##### Prioritizes Availability
+When Pebble encounters corruption, it returns an error for the affected keys in
+the SST, such as `pebble/table: invalid table 000004 (checksum mismatch at
+1615/1618)`. Despite this, the database remains available. Keys not affected by
+the corruption remain accessible, and new keys can be added to the database.
+Updating corrupted keys with the same data has no effect.
+
+#### RocksDB
+RocksDB is an embeddable persistent key-value store written in C++ and used by
+multiple databases for their physical durability layer.
+
+##### Truncates Faulty WAL Frames
+Like SQLite and Pebble, RocksDB truncates all subsequent WAL entries when it
+finds corruption in the WAL. Upon discovering corruption, it logs a warning,
+and the database remains available.
+
+```
+[db/db_impl/db_impl_open.cc:1119] 000004.log: dropping 1482 bytes; Corruption: checksum mismatch
+```
+
+##### Prioritizes Availability
+Similar to Pebble, RocksDB returns an error for keys in the SST affected by
+corruption, such as `Corruption: block checksum mismatch: stored(context
+removed) = 1930575595, computed = 3064321270, type = 4 in rocksdb/000008.sst
+offset 1613 size 1621`. Despite this, the database remains available. Keys not
+affected by the corruption remain accessible, and new keys can be added to the
+database.
+
+For my research into RocksDB and Pebble, See  https://github.com/thrawn01/db-corruption
+
+## Goals
+The goal of this proposal is to develop a robust and flexible error reporting
+and corruption handling system. This system will be suitable for building both
+durability and availability systems on top of SlateDB.
+
+- Ensure the database remains available in the face of corruption.
```

Comment:
What about manifest file corruption or missing SSTs?

Reply:
Excellent point, I will add notes for both manifest and missing SSTs